tower-http/src/classify.rs

```diff
@@ -0,0 +1,170 @@
+//! Tools for classifying responses as either success or failure.
+
+use http::{HeaderMap, Response, StatusCode};
+use http_body::Body;
+use std::{convert::Infallible, marker::PhantomData};
+
+/// Trait for classifying responses as either success or failure.
+pub trait ClassifyResponse {
+    /// The output of doing the classification.
+    ///
+    /// This could be an HTTP status code or some other kind of status.
+    type Output;
+
+    /// The type used to classify the end of streaming (Eos) responses.
+    type ClassifyEos: ClassifyEos<Output = Self::Output>;
+
+    /// Perform the classification.
+    fn classify_response<B>(
+        &mut self,
+        res: &Response<B>,
+    ) -> ClassifiedNowOrLater<Self::Output, Self::ClassifyEos>
+    where
+        B: Body;
+}
+
+/// Trait for classifying end of streams ("Eos") as either success or failure.
+pub trait ClassifyEos {
+    /// The output of doing the classification.
+    ///
+    /// This could be an HTTP status code or some other kind of status.
+    type Output;
+
+    /// Perform the classification.
+    fn classify_eos(&mut self, trailers: Option<&HeaderMap>) -> Classification<Self::Output>;
+}
+
+/// Result of doing a classification.
+#[derive(Debug)]
+// really don't like this name... :(
+pub enum ClassifiedNowOrLater<T, ClassifyEos> {
+    /// The response was able to be classified immediately.
+    Ready(Classification<T>),
+    /// We have to wait until the end of a streaming response to classify it.
+    RequiresEos(ClassifyEos),
+}
+
+/// A classification of a response.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Classification<T> {
+    /// The response was a success.
+    Success(T),
+    /// The response was a failure.
+    Failure(T),
+}
+
+/// A [`ClassifyEos`] type that can be used in [`ClassifyResponse`] implementations that never have
+/// to classify streaming responses.
+///
+/// `NeverClassifyEos` exists only as type. You cannot construct `NeverClassifyEos` values.
+pub struct NeverClassifyEos<T> {
+    _output_ty: PhantomData<T>,
+    _never: Infallible,
+}
+
+impl<T> ClassifyEos for NeverClassifyEos<T> {
+    type Output = T;
+
+    fn classify_eos(&mut self, _trailers: Option<&HeaderMap>) -> Classification<Self::Output> {
+        // `NeverClassifyEos` contains an `Infallible` so it can never be constructed
+        unreachable!()
+    }
+}
+
+/// The default classifier used for normal HTTP responses.
+///
+/// Responses with a `5xx` status code are failures, all others as successes.
+#[derive(Clone, Debug, Default)]
+pub struct ServerErrorsAsFailures {
+    _priv: (),
+}
```

Comment:
another premade classifier that might be nice to have is one that just takes a list of specific status codes to treat as failures, and treats everything else as a success, or one that takes a list of successes and treats everything else as failures. we can add that later, but i just thought i'd mention.

Reply:
Yeah that would be cool.